faunadb/expr.go

```diff
@@ -22,15 +27,112 @@ custom data structures. For example:
 */
 type Expr interface {
 	expr() // Make sure only internal structures can be marked as valid expressions
+	String() string
 }
 
 type unescapedObj map[string]Expr
+
+type rawForm []unescapedObj
+
+func (form rawForm) MarshalJSON() ([]byte, error) {
+	obj := unescapedObj{}
+	for _, entry := range form {
+		for k, v := range entry {
+			obj[k] = v
+		}
+	}
+	return json.Marshal(obj)
+}
```

Comment:
As a general feedback, I think this approach acquires too much runtime cost for the sake of string representability. If we can't derive the expression string representation from the structure, I'd say we're better off registering a function per expression that handles its string conversion rather than acquiring this extra CPU and memory pressure during the marshall path.

Reply:
I did a small research into it. I believe a path that allow us to encode more information into the function itself without penalizing runtime is to implement them as inner structs:

```go
func Abort(msg interface{}) Expr {
	type abort struct {
		fnApply
		Abort Expr `json:"abort"`
	}
	return abort{Abort: wrap(msg)}
}
```

`fnApply` implements `expr` and `String`. The string representation of the function call can be derived from the struct itself via reflection. On the other hand, serializing structs to json is faster than serializing maps, like we were doing before.

The missing part is how to handle optional arguments, though. I didn't have time to research that.